File: ports/jvm/src/main/java/com/unicodesecurity/FilenameDisguise.java
package com.unicodesecurity;

import java.util.ArrayList;
import java.util.List;

/**
 * FilenameDisguise — detection of filename/extension disguise attacks where the
 * visible extension differs from the byte extension (display-layer detector).
 *
 * <p>Byte-faithful transliteration of the verified Rust reference
 * implementation, itself a transliteration of
 * {@code Unicode/Security/Display/FilenameDisguise.lean}.
 *
 * <p>Threat model. An adversary delivers a file whose rendered name looks like a
 * benign type ({@code document.txt}) but whose actual byte extension is
 * executable — the canonical attack inserts {@code U+202E} RIGHT-TO-LEFT
 * OVERRIDE so {@code document<RLO>txt.exe} renders as {@code document exe.txt}.
 *
 * <p>Detection is presentation- and language-agnostic: it surfaces every
 * codepoint that could cause display-vs-byte divergence in the filename — any
 * bidi format-control anywhere, and any fullwidth/halfwidth or combining
 * (grapheme Extend) codepoint in the extension region (after the last
 * {@code .}). Native-RTL names with no bidi controls clear. It reuses the port's
 * own predicates (the bidi-format-control set, the grapheme Extend class, the
 * fullwidth range), never a host filesystem or rendering library.
 *
 * <p>Sub-threats (priority order):
 * <ol>
 *   <li>{@code RloFlip} — any bidi format-control in the input.</li>
 *   <li>{@code WidthClassExt} — a fullwidth/halfwidth codepoint in the
 *       extension.</li>
 *   <li>{@code CombiningInExt} — a combining (Extend) codepoint in the
 *       extension.</li>
 *   <li>{@code MultipleExtensions} — three or more dots (advisory; e.g.
 *       legitimate {@code .tar.gz.sig}).</li>
 * </ol>
 */
public final class FilenameDisguise {
  private FilenameDisguise() {}

  /** UTS #39-adjacent family key; the reason-code layer for this family is {@code D}. */
  public static final String FAMILY = "filename-disguise";

  // ───────────────────────────────────────────────────────────────────────
  // §1 Constants
  // ───────────────────────────────────────────────────────────────────────

  /** {@code U+002E FULL STOP} — the extension separator. */
  public static final int ASCII_DOT = 0x002E;

  /** The dot count at or beyond which the input is treated as a multi-extension advisory. */
  public static final int MIN_MULTIPLE_EXTENSIONS = 3;

  // ───────────────────────────────────────────────────────────────────────
  // §2 Types
  // ───────────────────────────────────────────────────────────────────────

  /** Sub-threat enumeration for FilenameDisguise, in priority order. */
  public sealed interface SubThreat
      permits RloFlip, WidthClassExt, CombiningInExt, MultipleExtensions {
    /** Fixture-row tag string for this sub-threat. */
    String tag();
  }

  /** A bidi format-control at {@code position} (codepoint {@code controlCp}). */
  public record RloFlip(int position, int controlCp) implements SubThreat {
    @Override
    public String tag() {
      return "RloFlip";
    }
  }

  /** A fullwidth/halfwidth codepoint in the extension, at {@code position} (codepoint {@code cp}). */
  public record WidthClassExt(int position, int cp) implements SubThreat {
    @Override
    public String tag() {
      return "WidthClassExt";
    }
  }

  /** A combining (grapheme Extend) codepoint in the extension, at {@code position} (codepoint {@code cp}). */
  public record CombiningInExt(int position, int cp) implements SubThreat {
    @Override
    public String tag() {
      return "CombiningInExt";
    }
  }

  /** Three or more {@code .} separators (advisory), {@code dotCount} of them. */
  public record MultipleExtensions(int dotCount) implements SubThreat {
    @Override
    public String tag() {
      return "MultipleExtensions";
    }
  }

  /** Top-level classification (safe = {@link Clear}). */
  public sealed interface Classification permits Clear, Hazard {
    /** True iff this classification is {@code Clear}. */
    boolean isClear();

    /** Human-facing tag for a hazard, or {@code null} when clear. */
    String tag();

    /** Implicated positions (empty when clear). */
    List<Integer> positions();
  }

  /** No disguise trigger present. */
  public record Clear() implements Classification {
    @Override
    public boolean isClear() {
      return true;
    }

    @Override
    public String tag() {
      return null;
    }

    @Override
    public List<Integer> positions() {
      return List.of();
    }
  }

  /** A disguise trigger fired. */
  public record Hazard(SubThreat sub, List<Integer> positions, List<Integer> decoded)
      implements Classification {
    @Override
    public boolean isClear() {
      return false;
    }

    @Override
    public String tag() {
      return sub.tag();
    }
  }

  /** The structured output of {@link #detect} (mirrors the Lean {@code Verdict}). */
  public record Verdict(
      List<Integer> input,
      Classification classify,
      List<Integer> dotPositions,
      Integer lastDotPos,
      int bidiControlCount,
      int fullwidthInExt,
      int combiningInExt) {}

  /**
   * The fixture reason code for a classification, or {@code null} when clear.
   * Mirrors the shared {@code unicode.security.D.filename-disguise.<tag>} scheme
   * keyed by the sub-threat tag.
   */
  public static String reasonCode(Classification classify) {
    String tag = classify.tag();
    return tag == null ? null : "unicode.security.D." + FAMILY + "." + tag;
  }

  // ───────────────────────────────────────────────────────────────────────
  // §3 Core predicates (each reuses the port's own table, never a host library)
  // ───────────────────────────────────────────────────────────────────────

  /** True iff {@code cp} is {@code U+002E FULL STOP} (the extension separator). */
  public static boolean isAsciiDot(int cp) {
    return cp == ASCII_DOT;
  }

  /** True iff {@code cp} is in the Halfwidth and Fullwidth Forms block. */
  public static boolean isFullwidthHalfwidth(int cp) {
    return cp >= 0xFF01 && cp <= 0xFFEF;
  }

  /** True iff {@code cp} is a bidi format-control (reuses {@link Security#isBidiFormatControl}). */
  public static boolean isBidiFormatControl(int cp) {
    return Security.isBidiFormatControl(cp);
  }

  /**
   * True iff {@code cp} has {@code Grapheme_Cluster_Break = Extend} (reuses the
   * port's own {@link Security#isGraphemeExtend}, derived from the bundled
   * Grapheme_Extend property unioned with the emoji modifiers).
   */
  public static boolean isGraphemeExtend(int cp) {
    return Security.isGraphemeExtend(cp);
  }

  // ───────────────────────────────────────────────────────────────────────
  // §4 Sub-detectors
  // ───────────────────────────────────────────────────────────────────────

  /** Positions of every {@code .} in {@code input}. */
  private static List<Integer> dotPositions(List<Integer> input) {
    List<Integer> out = new ArrayList<>();
    for (int i = 0; i < input.size(); i++) {
      if (isAsciiDot(input.get(i))) out.add(i);
    }
    return List.copyOf(out);
  }

  private static int countBidiControl(List<Integer> input) {
    int count = 0;
    for (int cp : input) {
      if (isBidiFormatControl(cp)) count++;
    }
    return count;
  }

  /** Position and codepoint of the first bidi format-control, or {@code null}. */
  private static int[] firstBidiControl(List<Integer> input) {
    for (int i = 0; i < input.size(); i++) {
      int cp = input.get(i);
      if (isBidiFormatControl(cp)) return new int[] {i, cp};
    }
    return null;
  }

  /** Position and codepoint of the first fullwidth/halfwidth codepoint at or after {@code start}, or {@code null}. */
  private static int[] firstFullwidthFrom(List<Integer> input, int start) {
    for (int i = start; i < input.size(); i++) {
      int cp = input.get(i);
      if (isFullwidthHalfwidth(cp)) return new int[] {i, cp};
    }
    return null;
  }

  /** Position and codepoint of the first Extend codepoint at or after {@code start}, or {@code null}. */
  private static int[] firstExtendFrom(List<Integer> input, int start) {
    for (int i = start; i < input.size(); i++) {
      int cp = input.get(i);
      if (isGraphemeExtend(cp)) return new int[] {i, cp};
    }
    return null;
  }

  /** Count of fullwidth/halfwidth codepoints at or after {@code start}. */
  private static int countFullwidthFrom(List<Integer> input, int start) {
    int count = 0;
    for (int i = start; i < input.size(); i++) {
      if (isFullwidthHalfwidth(input.get(i))) count++;
    }
    return count;
  }

  /** Count of Extend codepoints at or after {@code start}. */
  private static int countExtendFrom(List<Integer> input, int start) {
    int count = 0;
    for (int i = start; i < input.size(); i++) {
      if (isGraphemeExtend(input.get(i))) count++;
    }
    return count;
  }

  // ───────────────────────────────────────────────────────────────────────
  // §5 Top-level detection
  // ───────────────────────────────────────────────────────────────────────

  /** The FilenameDisguise detection function. */
  public static Verdict detect(List<Integer> input) {
    List<Integer> cps = List.copyOf(input);
    List<Integer> dots = dotPositions(cps);
    Integer lastDot = dots.isEmpty() ? null : dots.get(dots.size() - 1);
    int extStart = lastDot == null ? cps.size() : lastDot + 1;
    int bidiCount = countBidiControl(cps);
    int fwInExt = countFullwidthFrom(cps, extStart);
    int extInExt = countExtendFrom(cps, extStart);

    Classification classification = classify(cps, dots, extStart);

    return new Verdict(cps, classification, dots, lastDot, bidiCount, fwInExt, extInExt);
  }

  private static Classification classify(List<Integer> input, List<Integer> dots, int extStart) {
    // Priority 1: any bidi format-control anywhere in the input.
    int[] bidi = firstBidiControl(input);
    if (bidi != null) {
      return new Hazard(new RloFlip(bidi[0], bidi[1]), positionList(bidi[0]), List.of());
    }

    // Priority 2: a fullwidth/halfwidth codepoint in the extension region.
    int[] fw = firstFullwidthFrom(input, extStart);
    if (fw != null) {
      return new Hazard(new WidthClassExt(fw[0], fw[1]), positionList(fw[0]), List.of());
    }

    // Priority 3: a combining (Extend) codepoint in the extension region.
    int[] ext = firstExtendFrom(input, extStart);
    if (ext != null) {
      return new Hazard(new CombiningInExt(ext[0], ext[1]), positionList(ext[0]), List.of());
    }

    // Priority 4: three or more extensions (advisory).
    if (dots.size() >= MIN_MULTIPLE_EXTENSIONS) {
      return new Hazard(new MultipleExtensions(dots.size()), List.copyOf(dots), List.of());
    }

    return new Clear();
  }

  private static List<Integer> positionList(int position) {
    List<Integer> out = new ArrayList<>();
    out.add(position);
    return List.copyOf(out);
  }
}

File: ports/dotnet/src/UnicodeSecurity/Display/FilenameDisguise.cs
namespace UnicodeSecurity;

// filename-disguise (D layer) — detection of filename/extension disguise attacks
// where the visible extension differs from the byte extension (display-layer
// detector).
//
// Direct port of the verified Rust reference implementation, itself a
// byte-faithful transliteration of Unicode/Security/Display/FilenameDisguise.lean.
//
// Threat model. An adversary delivers a file whose rendered name looks like a
// benign type (document.txt) but whose actual byte extension is executable — the
// canonical attack inserts U+202E RIGHT-TO-LEFT OVERRIDE so document<RLO>txt.exe
// renders as document exe.txt.
//
// Detection is presentation- and language-agnostic: it surfaces every codepoint
// that could cause display-vs-byte divergence in the filename — any purposeless
// bidi format-control anywhere (BidiControlPurpose: unbalanced, or a balanced
// span enclosing nothing right-to-left in a left-to-right context), and any
// fullwidth/halfwidth or combining (grapheme Extend) codepoint in the extension
// region (after the last dot). Native-RTL names with no bidi controls clear, and
// so does a balanced embedding around an Arabic segment. It reuses the port's
// own predicates (the bidi purpose rule, the grapheme Extend class from the
// UAX #29 segmenter, and the fullwidth range), never a host filesystem or
// rendering library.
//
// Sub-threats (priority order):
//   1. RloFlip            a purposeless bidi format-control in the input.
//   2. WidthClassExt      a fullwidth/halfwidth codepoint in the extension.
//   3. CombiningInExt     a combining (Extend) codepoint in the extension.
//   4. MultipleExtensions >= 3 dots (advisory; e.g. legitimate .tar.gz.sig).
public static partial class Security
{
    public static class FilenameDisguise
    {
        // ─────────────────────────────────────────────────────────────────
        // §1 Types
        // ─────────────────────────────────────────────────────────────────

        /// <summary>Sub-threat enumeration for FilenameDisguise, in priority
        /// order. Each variant overrides <see cref="Tag"/>, so the fixture-row
        /// tag is resolved polymorphically with no catch-all.</summary>
        public abstract record SubThreat
        {
            /// <summary>Fixture-row tag string for this sub-threat (matches the
            /// Lean SubThreat.tag).</summary>
            public abstract string Tag { get; }
        }

        /// <summary>A bidi format-control at <see cref="Position"/> (codepoint
        /// <see cref="ControlCp"/>).</summary>
        public sealed record RloFlip(int Position, int ControlCp) : SubThreat
        {
            public override string Tag => "RloFlip";
        }

        /// <summary>A fullwidth/halfwidth codepoint in the extension at
        /// <see cref="Position"/> (codepoint <see cref="Cp"/>).</summary>
        public sealed record WidthClassExt(int Position, int Cp) : SubThreat
        {
            public override string Tag => "WidthClassExt";
        }

        /// <summary>A combining (grapheme Extend) codepoint in the extension at
        /// <see cref="Position"/> (codepoint <see cref="Cp"/>).</summary>
        public sealed record CombiningInExt(int Position, int Cp) : SubThreat
        {
            public override string Tag => "CombiningInExt";
        }

        /// <summary>Three or more dot separators (advisory).
        /// <see cref="DotCount"/> is the number of dots.</summary>
        public sealed record MultipleExtensions(int DotCount) : SubThreat
        {
            public override string Tag => "MultipleExtensions";
        }

        /// <summary>Top-level classification (no disguise trigger =
        /// <see cref="Clear"/>).</summary>
        public abstract record Classification
        {
            /// <summary>True iff the classification is Clear.</summary>
            public abstract bool IsClear { get; }

            /// <summary>Human-facing tag for a hazard, or null when clear.</summary>
            public abstract string? Tag { get; }

            /// <summary>Implicated positions (empty when clear).</summary>
            public abstract IReadOnlyList<int> Positions { get; }

            /// <summary>Fully-qualified reason code
            /// (unicode.security.D.filename-disguise.&lt;tag&gt;) routed through
            /// the port's shared reason-code machinery, or null when
            /// clear.</summary>
            public string? ReasonCode =>
                Tag is null ? null : Security.ReasonCode(Family.FilenameDisguise, Tag);
        }

        /// <summary>No disguise trigger present.</summary>
        public sealed record Clear : Classification
        {
            public override bool IsClear => true;
            public override string? Tag => null;
            public override IReadOnlyList<int> Positions => System.Array.Empty<int>();
        }

        /// <summary>A disguise trigger fired: the sub-threat, the implicated
        /// positions, and the (always-empty for this detector) decoded-byte
        /// projection, kept for shape parity with the Lean
        /// Classification.hazard.</summary>
        public sealed record Hazard(
            SubThreat Sub,
            IReadOnlyList<int> HazardPositions,
            IReadOnlyList<int> Decoded) : Classification
        {
            public override bool IsClear => false;
            public override string? Tag => Sub.Tag;
            public override IReadOnlyList<int> Positions => HazardPositions;
        }

        /// <summary>The structured output of <see cref="Detect"/> (mirrors the
        /// Lean Verdict).</summary>
        public sealed record Verdict(
            IReadOnlyList<int> Input,
            Classification Classify,
            IReadOnlyList<int> DotPositions,
            int? LastDotPos,
            int BidiControlCount,
            int FullwidthInExt,
            int CombiningInExtCount);

        // ─────────────────────────────────────────────────────────────────
        // §2 Core predicates
        // ─────────────────────────────────────────────────────────────────

        /// <summary>True iff <paramref name="cp"/> is U+002E FULL STOP (the
        /// extension separator).</summary>
        public static bool IsAsciiDot(int cp) => cp == 0x002E;

        /// <summary>True iff <paramref name="cp"/> is in the Halfwidth and
        /// Fullwidth Forms block (reuses the homoglyph detector's
        /// IsFullwidthHalfwidth, which is exactly U+FF01..U+FFEF).</summary>
        public static bool IsFullwidthHalfwidth(int cp) => Security.IsFullwidthHalfwidth(cp);

        /// <summary>True iff <paramref name="cp"/> is a bidi format-control
        /// (reuses the port's own IsBidiFormatControl predicate — the
        /// LRE/RLE/LRO/RLO/PDF/LRI/RLI/FSI/PDI set).</summary>
        public static bool IsBidiFormatControl(int cp) => Security.IsBidiFormatControl(cp);

        /// <summary>True iff <paramref name="cp"/> has
        /// Grapheme_Cluster_Break = Extend (reuses the UAX #29 segmenter's GCB
        /// table).</summary>
        public static bool IsGraphemeExtend(int cp) =>
            Segmentation.Grapheme.LookupGcb((uint)cp) == Segmentation.Gcb.Extend;

        // ─────────────────────────────────────────────────────────────────
        // §3 Sub-detectors
        // ─────────────────────────────────────────────────────────────────

        /// <summary>Positions of every dot in <paramref name="input"/>.</summary>
        private static List<int> DotPositions(IReadOnlyList<int> input)
        {
            var positions = new List<int>();
            for (var idx = 0; idx < input.Count; idx++)
            {
                if (IsAsciiDot(input[idx])) positions.Add(idx);
            }
            return positions;
        }

        /// <summary>Position and codepoint of the first purposeless bidi
        /// format-control, or null: unbalanced, or a balanced span enclosing
        /// nothing right-to-left in a left-to-right context
        /// (BidiControlPurpose.FirstPurposelessControl). A balanced embedding
        /// around an Arabic filename segment manages that segment and is not a
        /// flip. Mirrors the Lean detect, which reads
        /// firstPurposelessControl.</summary>
        private static (int Pos, int Cp)? FirstBidiControl(IReadOnlyList<int> input) =>
            BidiControlPurpose.FirstPurposelessControl(input);

        /// <summary>Position and codepoint of the first fullwidth/halfwidth
        /// codepoint at or after <paramref name="start"/>, or null when none is
        /// present.</summary>
        private static (int Pos, int Cp)? FirstFullwidthFrom(IReadOnlyList<int> input, int start)
        {
            for (var idx = 0; idx < input.Count; idx++)
            {
                if (idx >= start && IsFullwidthHalfwidth(input[idx])) return (idx, input[idx]);
            }
            return null;
        }

        /// <summary>Position and codepoint of the first Extend codepoint at or
        /// after <paramref name="start"/>, or null when none is present.</summary>
        private static (int Pos, int Cp)? FirstExtendFrom(IReadOnlyList<int> input, int start)
        {
            for (var idx = 0; idx < input.Count; idx++)
            {
                if (idx >= start && IsGraphemeExtend(input[idx])) return (idx, input[idx]);
            }
            return null;
        }

        /// <summary>Count of fullwidth/halfwidth codepoints at or after
        /// <paramref name="start"/>.</summary>
        private static int CountFullwidthFrom(IReadOnlyList<int> input, int start)
        {
            var count = 0;
            for (var idx = 0; idx < input.Count; idx++)
            {
                if (idx >= start && IsFullwidthHalfwidth(input[idx])) count++;
            }
            return count;
        }

        /// <summary>Count of Extend codepoints at or after
        /// <paramref name="start"/>.</summary>
        private static int CountExtendFrom(IReadOnlyList<int> input, int start)
        {
            var count = 0;
            for (var idx = 0; idx < input.Count; idx++)
            {
                if (idx >= start && IsGraphemeExtend(input[idx])) count++;
            }
            return count;
        }

        private static int CountBidiControl(IReadOnlyList<int> input)
        {
            var count = 0;
            foreach (var cp in input)
            {
                if (IsBidiFormatControl(cp)) count++;
            }
            return count;
        }

        // ─────────────────────────────────────────────────────────────────
        // §4 Top-level detection
        // ─────────────────────────────────────────────────────────────────

        /// <summary>The FilenameDisguise detection function, reading its input
        /// as one filename. Mirrors the Lean detect, which is detectWithContext
        /// at the default context.</summary>
        public static Verdict Detect(IReadOnlyList<int> input) => DetectWithContext(false, input);

        /// <summary>The FilenameDisguise detection function under an explicit
        /// field context. <paramref name="runningText"/> mirrors the Lean
        /// Context.runningText: the extension rungs read the text after the last
        /// dot as a file extension, which a source file or a message does not
        /// have, so they do not run on running text; the purposeless-bidi-control
        /// rung holds of any field.</summary>
        public static Verdict DetectWithContext(bool runningText, IReadOnlyList<int> input)
        {
            var dots = DotPositions(input);
            int? lastDot = dots.Count == 0 ? null : dots[dots.Count - 1];
            var extStart = lastDot is int p ? p + 1 : input.Count;
            var bidiCount = CountBidiControl(input);
            var fwInExt = CountFullwidthFrom(input, extStart);
            var extInExt = CountExtendFrom(input, extStart);

            Classification classification;
            // Priority 1: a purposeless bidi format-control anywhere in the input.
            if (FirstBidiControl(input) is (int ctlPos, int ctlCp))
            {
                classification = new Hazard(
                    new RloFlip(ctlPos, ctlCp),
                    new List<int> { ctlPos },
                    new List<int>());
            }
            // The remaining rungs read an extension; running text has none.
            else if (runningText)
            {
                classification = new Clear();
            }
            // Priority 2: fullwidth/halfwidth in the extension.
            else if (FirstFullwidthFrom(input, extStart) is (int fwPos, int fwCp))
            {
                classification = new Hazard(
                    new WidthClassExt(fwPos, fwCp),
                    new List<int> { fwPos },
                    new List<int>());
            }
            // Priority 3: combining mark in the extension.
            else if (FirstExtendFrom(input, extStart) is (int exPos, int exCp))
            {
                classification = new Hazard(
                    new CombiningInExt(exPos, exCp),
                    new List<int> { exPos },
                    new List<int>());
            }
            // Priority 4: three or more extensions (advisory).
            else if (dots.Count >= 3)
            {
                classification = new Hazard(
                    new MultipleExtensions(dots.Count),
                    new List<int>(dots),
                    new List<int>());
            }
            else
            {
                classification = new Clear();
            }

            return new Verdict(
                input.ToList(),
                classification,
                dots,
                lastDot,
                bidiCount,
                fwInExt,
                extInExt);
        }
    }
}
